[c10d][gloo] Integrate vendor generic FR into gloo

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #152614
* #154929

This is a first quick prototyping for FR integration for gloo. Few features gaps:
- Input/Output numels for each collective
- Whether to use c10::Event or where to use it.
- Where to dump the FR traces. (The dump api is provided in this PR)


cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k

Differential Revision: [D75803601](https://our.internmc.facebook.com/intern/diff/D75803601)

fduwjj committed:
[WIP] Make FR vendor generic and try to enable it for gloo

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Make FR vendor generic and try to enable it for gloo"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Make FR vendor generic and try to enable it for gloo"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Make FR vendor generic and try to enable it for gloo"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Make FR vendor generic and try to enable it for gloo"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

d4l3k commented on torch/csrc/distributed/c10d/ProcessGroupGloo.cpp:
```diff
@@ -1884,6 +1969,18 @@ class AsyncAllgatherWork : public ProcessGroupGloo::AsyncWork {
 
+  std::chrono::milliseconds getTimeout() const override {
+    return context->getTimeout();
+  }
```
can we move this into AsyncWork parent?

fduwjj replied:
sure.

d4l3k replied:
still relevant

d4l3k reviewed (commented):
I think this is a very reasonable way to integrate into Gloo

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupGloo.cpp:
```diff
@@ -879,15 +888,43 @@ void ProcessGroupGloo::runLoop(int workerIndex) {
   workQueue_.push_back(std::move(work));
   lock.unlock();
+  // using c10d::FlightRecorder;
+  work->trace_id_ = FlightRecorder::get()->record(
```
this part looks right to me. (straightforward..) - recording event at the time of dispatching the async kernel.

however what i'm not clear on is how gloo implements timeouts.  Does it even have an equivalent of a watchdog thread, or is it going to wait until someone checks the result of a future to finally raise the timeout exception?

wconstab replied:
actually, is it safer to record the FR event _before_ acquiring the workMutex?

i'm not sure under what circumstances gloo could deadlock, but the question is whether we'd want to have a record in FR for the about-to-enqueue work that didn't enqueue due to a deadlock

fduwjj replied:
It is using a side thread to run the collective as well.. But we just use that thread to call exec() and IIUC lots of collectives are blocking so there is no timeout check even. By tracing the code inside gloo, we do the waitSend/waitRecv with timeout inside the implementation of gloo. So there is no way we can do the timeout check in c10d layer, at least for now.

And for recording, I prefer to record it when everything is settled. Because since gloo is synchronous we know which collective will timeout exactly.

cyyever commented:
For the failures, we can now use `std::filesystem::create_directories`

fduwjj committed:
Update on "[WIP] Make FR vendor generic and try to enable it for gloo"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Make FR vendor generic and try to enable it for gloo"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Make FR vendor generic and try to enable it for gloo"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Make FR vendor generic and try to enable it for gloo"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

d4l3k commented on torch/csrc/distributed/c10d/ProcessGroupGloo.cpp:
```diff
@@ -879,15 +888,43 @@ void ProcessGroupGloo::runLoop(int workerIndex) {
 
     AsyncWork::execute(work);
+    // TODO: Need to find a way to calculate the difference of duration of two
+    // c10d::Event
```
`AsyncWork::execute(work)` is a blocking call and returns once the work has completed. We should be able to just measure the wallclock time for CPU operations.

For GPU operations the CUDAWork objects have a `.synchronize()` in them so the timing will be imperfect unless we track cuda events for those.

I.e. https://github.com/pytorch/pytorch/blob/e6ef0620ccaba6f00ab39d652b451835800d8c10/torch/csrc/distributed/c10d/ProcessGroupGloo.cpp#L954

I wonder if we should make an AsyncCPUWork that measures wall clock and AsyncCUDAWork which only starts the clock after the synchronize. We could also just track CUDAWork using CUDAEvents but not strictly necessary

d4l3k reviewed (approved):
Code looks reasonable for the initial integration other than the small amount of cleanup on the duplicated methods

fduwjj committed:
Update on "[WIP] Make FR vendor generic and try to enable it for gloo"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Make FR vendor generic and try to enable it for gloo"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Make FR vendor generic and try to enable it for gloo"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Make FR vendor generic and try to enable it for gloo"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented:
@d4l3k Still not aiming to land now, we need the following:
1. A dump API for gloo, like PGNCCL.
2. Unit test.
3. A gating for the FR for gloo. Currently since FR is by default turned on for everyone.

fduwjj committed:
Update on "[WIP] Make FR vendor generic and try to enable it for gloo"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Make FR vendor generic and try to enable it for gloo"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][gloo] Integrate vendor generic FR into gloo"


This is a first quick prototyping for FR integration for gloo. Few features gaps:
- Input/Output numels for each collective
- Whether to use c10::Event or where to use it.
- Where to dump the FR traces. (The dump api is provided in this PR)


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][gloo] Integrate vendor generic FR into gloo"


This is a first quick prototyping for FR integration for gloo. Few features gaps:
- Input/Output numels for each collective
- Whether to use c10::Event or where to use it.
- Where to dump the FR traces. (The dump api is provided in this PR)


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][gloo] Integrate vendor generic FR into gloo"


This is a first quick prototyping for FR integration for gloo. Few features gaps:
- Input/Output numels for each collective
- Whether to use c10::Event or where to use it.
- Where to dump the FR traces. (The dump api is provided in this PR)


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][gloo] Integrate vendor generic FR into gloo"


This is a first quick prototyping for FR integration for gloo. Few features gaps:
- Input/Output numels for each collective
- Whether to use c10::Event or where to use it.
- Where to dump the FR traces. (The dump api is provided in this PR)


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D75803601).

fduwjj committed:
Update on "[c10d][gloo] Integrate vendor generic FR into gloo"


This is a first quick prototyping for FR integration for gloo. Few features gaps:
- Input/Output numels for each collective
- Whether to use c10::Event or where to use it.
- Where to dump the FR traces. (The dump api is provided in this PR)


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

Differential Revision: [D75803601](https://our.internmc.facebook.com/intern/diff/D75803601)

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][gloo] Integrate vendor generic FR into gloo"


This is a first quick prototyping for FR integration for gloo. Few features gaps:
- Input/Output numels for each collective
- Whether to use c10::Event or where to use it.
- Where to dump the FR traces. (The dump api is provided in this PR)


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

Differential Revision: [D75803601](https://our.internmc.facebook.com/intern/diff/D75803601)

[ghstack-poisoned]

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D75803601).

fduwjj commented:
@pytorchbot merge -i
